### ISSUE TITLE
short-circuit AST visits in registerNodeProcessors(where possible)

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#2259]
See for example #2258 which demonstrates how error-prone our current approach is.  Ideally we wouldn't be visiting at all in this case and the visitor not registered in `registerNodeProcessors`.

**Comment by pq** [comment]
**Forward-pointer.**  Consider something like:

```dart

  @override
  void registerNodeProcessors(
      NodeLintRegistry registry, LinterContext context) {
    if (!isInLibDir(context.currentUnit.unit, context.package)) {
      return;
    }
   ...
}

```

(Untested.)

**Comment by pq** [comment]
Fixed w/ #2263, #2264, and #2265.
